### DISCUSSION TITLE
Use mask instead of polygons for training.

### DISCUSSION THREAD

**Discussion by 7f0c** [ultralytics/ultralytics#6721] [Q&A]
Is this possible to do? My project involves objects that are a bit tedious to annotate with polygons but quite easy when using masks instead.

**Comment by pderrenger** [comment]
@7f0c yes, it's possible to train YOLOv8 models using masks for segmentation tasks. YOLOv8 supports training with masks, which can be more convenient for certain types of objects where creating masks is easier than annotating with polygons. Make sure your dataset is prepared with the appropriate mask annotations, and you can proceed with training your model for segmentation. For more details on how to set up your training with masks, please refer to the [Segmentation Task Documentation](https://docs.ultralytics.com/tasks/segment). 😊👍

**Reply by 7f0c** [reply]
I was thinking of maybe using image masks instead of polygons. Something like this: 

![image_part_008](https://github.com/ultralytics/ultralytics/assets/130174233/85ffe8fb-cd2c-4f7b-a19d-462d2cdadcc5)
![image_part_008](https://github.com/ultralytics/ultralytics/assets/130174233/14f1b1bc-5509-4b48-b1c9-c6f327d48e1c)

The [documentation](https://docs.ultralytics.com/datasets/segment/#ultralytics-yolo-format) says to use bounding coordinates to form the mask but my dataset is not suitable for that.

**Reply by glenn-jocher** [reply]
@7f0c looks good, but you'd need to convert your mask image to the required segmentation dataset polygon format first before attempting to train.

**Reply by 7f0c** [reply]
It's a bit tedious though to convert it to polygons. Can you point the way to change YOLO's dataloader so that it uses masks instead of polygons, if that's possible.

**Reply by glenn-jocher** [reply]
@7f0c absolutely, I understand that converting to polygons can be a bit of a hassle for certain projects. 😅 While YOLOv8 primarily uses polygons for segmentation tasks, adapting it to work directly with masks is possible but requires a bit of customization.

You'd need to modify the dataset loading part of the code to accept mask images directly. This involves adjusting the `load_mosaic()` and `load_image()` functions within the dataset class (`datasets.py`), where you'd integrate your mask loading logic. Here's a very simplified snippet to give you an idea:

```python
import cv2
import numpy as np

def load_mask(mask_path):
    # Example function to load mask
    mask = cv2.imread(mask_path, cv2.IMREAD_GRAYSCALE)  # Load mask
    return mask

class CustomDataset(Dataset):
    # Your custom dataset class
    def __getitem__(self, index):
        # Load your mask here instead of polygons
        mask = load_mask(self.mask_paths[index])
        # Convert your mask to the desired format here
        # This part depends on how you plan to use the masks with YOLOv8
```

Remember, this is a simplified example. You'll need to integrate it properly with the existing YOLOv8 dataloader logic, ensuring masks are correctly processed and fed into the model for training. This might involve additional steps like resizing masks to match the input size of the model and converting binary masks to a format that YOLOv8 can use for segmentation (e.g., creating segmentation maps or converting masks to bounding box + polygon format internally).

Hope this points you in the right direction! 🚀
